pkg/chunk/cache/redis_client.go

```diff
@@ -0,0 +1,304 @@
+		rdb:        redis.NewClient(opt),
+	}
+}
+
+func newRedisClusterClient(cfg *RedisConfig) *redisBasicClient {
+	opt := &redis.ClusterOptions{
+		Addrs:       strings.Split(cfg.Endpoint, ","),
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisBasicClient{
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewClusterClient(opt),
+	}
+}
+
+func (c *redisBasicClient) Ping(ctx context.Context) error {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	pong, err := c.rdb.Ping(ctx).Result()
+	if err != nil {
+		return err
+	}
+	if pong != "PONG" {
+		return fmt.Errorf("redis: Unexpected PING response %q", pong)
+	}
+	return nil
+}
+
+func (c *redisBasicClient) MSet(ctx context.Context, keys []string, values [][]byte) error {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	pipe := c.rdb.TxPipeline()
+	for i := range keys {
+		pipe.Set(ctx, keys[i], values[i], c.expiration)
+	}
+	_, err := pipe.Exec(ctx)
+	return err
+}
+
+func (c *redisBasicClient) MGet(ctx context.Context, keys []string) ([][]byte, error) {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	cmd := c.rdb.MGet(ctx, keys...)
+	if err := cmd.Err(); err != nil {
+		return nil, err
+	}
+
+	ret := make([][]byte, len(keys))
+	for i, val := range cmd.Val() {
+		if val != nil {
+			ret[i] = StringToBytes(val.(string))
+		}
+	}
+	return ret, nil
+}
+
+func (c *redisBasicClient) Close() error {
+	return c.rdb.Close()
+}
+
+type redisSentinelClient struct {
+	expiration time.Duration
+	timeout    time.Duration
+	rdb        *redis.SentinelClient
+	opt        *redis.Options
+}
+
+func newRedisSentinelClient(cfg *RedisConfig) *redisSentinelClient {
+	opt := &redis.Options{
+		Addr:        cfg.Endpoint,
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisSentinelClient{
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewSentinelClient(opt),
+		opt:        opt,
+	}
+}
+
+func (c *redisSentinelClient) getMaster(ctx context.Context) (*redisBasicClient, error) {
+	masters, err := c.rdb.Masters(ctx).Result()
+	if err != nil {
+		return nil, err
+	}
+	if len(masters) == 0 {
+		return nil, ErrNoMasters
+	}
+	for _, master := range masters {
+		// expected: master = []interface {}{"name", "<master name>", "ip", "<IP>", "port", "<port>", ... }
+		if info, ok := master.([]interface{}); ok {
+			if len(info) >= 6 && info[2] == "ip" && info[4] == "port" {
+				return &redisBasicClient{
```

Comment:
Does this mean we create a new TCP connection for every single `MSet()` and `MGet()` operation?

Reply:
AFAIK this is how sentinel works. First we are getting one of the masters, then we redirect the request to it.

Reply:
I'm not familiar with Redis sentinel (never used it), but I believe it's not feasible to open a new TCP connection for every single cache operation. This will not scale (other then being slow).